src/libcollections/range.rs

```diff
@@ -23,39 +23,75 @@ pub trait RangeArgument<T> {
     /// Start index (inclusive)
     ///
     /// Return start value if present, else `None`.
-    fn start(&self) -> Option<&T> {
+    fn start(&self) -> Option<T> {
         None
     }
 
     /// End index (exclusive)
     ///
     /// Return end value if present, else `None`.
-    fn end(&self) -> Option<&T> {
+    fn end(&self) -> Option<T> {
         None
     }
 }
 
-// FIXME add inclusive ranges to RangeArgument
-
 impl<T> RangeArgument<T> for RangeFull {}
 
-impl<T> RangeArgument<T> for RangeFrom<T> {
-    fn start(&self) -> Option<&T> {
-        Some(&self.start)
+impl<T: Copy> RangeArgument<T> for RangeFrom<T> {
+    fn start(&self) -> Option<T> {
+        Some(self.start)
     }
 }
 
-impl<T> RangeArgument<T> for RangeTo<T> {
-    fn end(&self) -> Option<&T> {
-        Some(&self.end)
+impl<T: Copy> RangeArgument<T> for RangeTo<T> {
+    fn end(&self) -> Option<T> {
+        Some(self.end)
     }
 }
 
-impl<T> RangeArgument<T> for Range<T> {
-    fn start(&self) -> Option<&T> {
-        Some(&self.start)
+impl<T: Copy> RangeArgument<T> for Range<T> {
+    fn start(&self) -> Option<T> {
+        Some(self.start)
+    }
+    fn end(&self) -> Option<T> {
+        Some(self.end)
     }
-    fn end(&self) -> Option<&T> {
-        Some(&self.end)
+}
+
+macro_rules! inclusive {
+    ($Int: ty) => {
+        impl RangeArgument<$Int> for RangeToInclusive<$Int> {
+            fn end(&self) -> Option<$Int> {
+                Some(self.end.checked_add(1).expect("inclusive range to maximum usize"))
+            }
+        }
+
+        impl RangeArgument<$Int> for RangeInclusive<$Int> {
+            fn start(&self) -> Option<$Int> {
+                match *self {
+                    RangeInclusive::Empty { at } => Some(at),
+                    RangeInclusive::NonEmpty { start, .. } => Some(start),
+                }
+            }
+            fn end(&self) -> Option<$Int> {
+                match *self {
+                    RangeInclusive::Empty { at } => Some(at),
+                    RangeInclusive::NonEmpty { end, .. } => {
+                        Some(end.checked_add(1).expect("inclusive range to maximum usize"))
```

Comment:
The message is wrong, isn't it? When this is expanded for `u8`, `(0...255).end()` will panic here. Same for the one above.